src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -184,95 +184,112 @@ bool tryBindAsBuilderType([NotNullWhen(true)] out BoundInterpolatedString? resul
-                // is an applicable AppendFormatted method that accepts them. To preserve langversion, we therefore make sure all components
-                // are convertible to object if the current langversion is lower than the interpolation feature
-                TypeSymbol? objectType = null;
-                BindingDiagnosticBag? conversionDiagnostics = null;
-                var needToCheckConversionToObject = !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics;
-                if (needToCheckConversionToObject)
-                {
-                    objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
-                    conversionDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false);
-                }
+        private BoundInterpolatedString BindUnconvertedInterpolatedStringToHandlerType(BoundUnconvertedInterpolatedString unconvertedInterpolatedString, NamedTypeSymbol interpolatedStringHandlerType, BindingDiagnosticBag diagnostics, bool isHandlerConversion)
+        {
+            diagnostics.Add(interpolatedStringHandlerType.GetUseSiteInfo(), unconvertedInterpolatedString.Syntax.Location);
+
+            // We satisfy the conditions for using an interpolated string builder. Bind all the builder calls unconditionally, so that if
+            // there are errors we get better diagnostics than "could not convert to object."
+            var (appendCalls, usesBoolReturn) = BindInterpolatedStringAppendCalls(unconvertedInterpolatedString, interpolatedStringHandlerType, diagnostics);
+
+            // Prior to C# 10, all types in an interpolated string expression needed to be convertible to `object`. After 10, some types
+            // (such as Span<T>) that are not convertible to `object` are permissible as interpolated string components, provided there
+            // is an applicable AppendFormatted method that accepts them. To preserve langversion, we therefore make sure all components
+            // are convertible to object if the current langversion is lower than the interpolation feature and we're converting this
+            // interpolation into an actual string.
+            TypeSymbol? objectType = null;
+            BindingDiagnosticBag? conversionDiagnostics = null;
+            var needToCheckConversionToObject = !isHandlerConversion && !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics;
+            if (needToCheckConversionToObject)
+            {
+                objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                conversionDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false);
+            }
+            else if (isHandlerConversion)
```

Comment:
>else if (isHandlerConversion) [](http://example.com/codeflow?start=12&length=29)

Consider making this a separate ```if``` rather than an ```else if```. This will avoid dependency on value of needToCheckConversionToObject and any future changes for it. ```needToCheckConversionToObject``` depends on ```isHandlerConversion```, not the other way around.